src/query/api/v1/handler/ready.go

```diff
@@ -0,0 +1,186 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package handler
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/m3db/m3/src/query/api/v1/options"
+	"github.com/m3db/m3/src/query/storage/m3"
+	"github.com/m3db/m3/src/query/util/logging"
+	xerrors "github.com/m3db/m3/src/x/errors"
+	"github.com/m3db/m3/src/x/instrument"
+	xhttp "github.com/m3db/m3/src/x/net/http"
+
+	"go.uber.org/zap"
+)
+
+const (
+	// ReadyURL is the url to check for readyiness.
+	ReadyURL = "/ready"
+
+	// ReadyHTTPMethod is the HTTP method used with this resource.
+	ReadyHTTPMethod = http.MethodGet
+)
+
+// ReadyHandler tests whether the service is connected to underlying storage.
+type ReadyHandler struct {
+	clusters       m3.Clusters
+	instrumentOpts instrument.Options
+}
+
+// NewReadyHandler returns a new instance of handler.
+func NewReadyHandler(opts options.HandlerOptions) http.Handler {
+	return &ReadyHandler{
+		clusters:       opts.Clusters(),
+		instrumentOpts: opts.InstrumentOpts(),
+	}
+}
+
+type readyResultNamespace struct {
+	ID         string                         `json:"id"`
+	Attributes readyResultNamespaceAttributes `json:"attributes"`
+}
+
+type readyResultNamespaceAttributes struct {
+	MetricsType string `json:"metricsType"`
+	Retention   string `json:"retention"`
+	Resolution  string `json:"resolution"`
+}
+
+type readyResult struct {
+	ReadyReads     []readyResultNamespace `json:"readyReads,omitempty"`
+	NotReadyReads  []readyResultNamespace `json:"notReadyReads,omitempty"`
+	ReadyWrites    []readyResultNamespace `json:"readyWrites,omitempty"`
+	NotReadyWrites []readyResultNamespace `json:"notReadyWrites,omitempty"`
+}
+
+func (h *ReadyHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	logger := logging.WithContext(r.Context(), h.instrumentOpts)
+
+	req, err := h.parseReadyRequestChecked(r)
+	if err != nil {
+		logger.Error("unable to parse ready request", zap.Error(err))
+		xhttp.WriteError(w, err)
+		return
+	}
+
+	var (
+		namespaces = h.clusters.ClusterNamespaces()
+		result     = &readyResult{}
+	)
+	for _, ns := range namespaces {
+		attrs := ns.Options().Attributes()
+		nsResult := readyResultNamespace{
+			ID: ns.NamespaceID().String(),
+			Attributes: readyResultNamespaceAttributes{
+				MetricsType: attrs.MetricsType.String(),
+				Retention:   attrs.Retention.String(),
+				Resolution:  attrs.Resolution.String(),
+			},
+		}
+
+		ready, err := ns.Session().ReadClusterAvailability()
+		if err != nil {
+			logger.Error("check read availability error", zap.Error(err))
+			xhttp.WriteError(w, err)
+			return
+		}
+		if !ready {
+			result.NotReadyReads = append(result.NotReadyReads, nsResult)
+		} else {
+			result.ReadyReads = append(result.ReadyReads, nsResult)
+		}
+
+		ready, err = ns.Session().WriteClusterAvailability()
```

Comment:
wdyt about skipping the respective availability check if not requested? e.g skip checking write available if req.writes is false. my thinking is it would be unfortunate to block reads if only reads were requested and the write availability ran into an error.

Reply:
So these don't actually wait on anything it's just a check of current state (checking open connection count), not actually trying to do reachability so won't be any faster/slower or blocking if one is unhealthy vs another.